File: src/main/scala/kafkastreams/scalaexamples/BranchExample.scala
package kafkastreams.scalaexamples

import com.fasterxml.jackson.databind.JsonNode
import kafkastreams.scalautils.JacksonDSL._
import kafkastreams.serdes.JsonNodeSerde
import org.apache.kafka.streams.Topology
import org.apache.kafka.streams.scala.ImplicitConversions._
import org.apache.kafka.streams.scala.Serdes.String
import org.apache.kafka.streams.scala.StreamsBuilder

class BranchExample extends KafkaStreamsApp {

  def createTopology(builder: StreamsBuilder): Topology = {
    implicit val json = new JsonNodeSerde

    val articles = builder.stream[String, JsonNode]("Articles")

    val articlesPerSite = articles.branch(
      (key, article) => article("site").asText == "bbc",
      (key, article) => article("site").asText == "cnn",
      (key, article) => article("site").asText == "foxnews",
      (key, article) => true // catch remaining events
    )

    val topics = List("BBC-Articles", "CNN-Articles", "FoxNews-Articles", "Other-Articles")

    for ((stream, topic) <- articlesPerSite.zip(topics))
      stream.to(topic)

    /* Alternatively, using new 'to' method:

    articles.to((key, article, recordContext) =>
      article("site").asText match {
        case "bbc" => "BBC-Articles"
        case "cnn" => "CNN-Articles"
        case "foxnews" => "FoxNews-Articles"
        case _ => "Other-Articles"
      }
    )
     */

    builder.build()
  }
}

object BranchExample extends App {
  new BranchExample().start("branch-app")
}

File: src/main/java/kafkastreams/javaexamples/BranchExample.java
package kafkastreams.javaexamples;

import com.fasterxml.jackson.databind.JsonNode;
import kafkastreams.serdes.JsonNodeSerde;
import org.apache.kafka.common.serialization.Serde;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.Topology;
import org.apache.kafka.streams.kstream.Consumed;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.Produced;

public class BranchExample extends KafkaStreamsApp {

    public static void main(String[] args) {
        new BranchExample().start("branch-app");
    }

    public Topology createTopology(StreamsBuilder builder) {
        Serde<String> strings = Serdes.String();
        JsonNodeSerde json = new JsonNodeSerde();

        KStream<String, JsonNode> articles = builder.stream("Articles", Consumed.with(strings, json));

        KStream<String, JsonNode>[] articlesPerSite = articles.branch(
                (key, value) -> "bbc".equals(value.path("site").asText()),
                (key, value) -> "cnn".equals(value.path("site").asText()),
                (key, value) -> "foxnews".equals(value.path("site").asText()),
                (key, value) -> true // catch remaining events
        );

        articlesPerSite[0].to("BBC-Articles", Produced.with(strings, json));
        articlesPerSite[1].to("CNN-Articles", Produced.with(strings, json));
        articlesPerSite[2].to("FoxNews-Articles", Produced.with(strings, json));
        articlesPerSite[3].to("Other-Articles", Produced.with(strings, json));

        /* Alternatively, using new 'to' method:

        articles.to((key, value, recordContext) -> {
            switch (value.path("site").asText()) {
                case "bbc": return "BBC-Articles";
                case "cnn": return "BBC-Articles";
                case "foxnews": return "FoxNews-Articles";
                default: return "Other-Articles";
            }
        });
         */

        return builder.build();
    }
}
